src/hevm/src/EVM.hs

```diff
@@ -1951,7 +1961,30 @@ type CheatAction = Word -> Word -> Buffer -> EVM ()
 cheatActions :: Map Word32 CheatAction
 cheatActions =
   Map.fromList
-    [ action "warp(uint256)" $
+    [ action "ffi(string[])" $
+        \sig outOffset outSize input -> do
+          vm <- get
+          if vmoptAllowFFI (view EVM.options vm) then
+            case decodeBuffer [AbiArrayDynamicType AbiStringType] input of
+              CAbi valsArr -> case valsArr of
+                [AbiArrayDynamic AbiStringType strsV] ->
+                  let
+                    cmd = (flip fmap) (V.toList strsV) (\case
+                      (AbiString a) -> unpack $ decodeUtf8 a
+                      _ -> "")
+                    cont = \bs -> do
+                      let encoded = ConcreteBuffer bs
+                      assign (state . returndata) encoded
+                      copyBytesToMemory encoded outSize 0 outOffset
+                      assign result Nothing
+                  in assign result (Just . VMFailure . Query $ (PleaseDoFFI cmd cont))
+                _ -> vmError (BadCheatCode sig)
+              _ -> vmError (BadCheatCode sig)
+          else
+            let msg = "ffi disabled: run again with --ffi if you want to allow tests to call external scripts"
+            in vmError . Revert $ abiMethod "Error(string)" (AbiString msg),
```

Comment:
I can't seem to get this message to pretty print in the trace, not sure what I'm doing wrong :/

Reply:
cheat codes aren't handled exactly like a call is, there is no new frame allocated and we don't call `finishFrame` when returning. The revert message is added to the trace in the `finalize` subcall of `finishFrame`

Reply:
no I mean the revert is shown in the trace, but it's just a bunch of hex data instead of being pretty printed as a string.

I see this:

```
Failure: testFFI
  src/broked.sol:CheatCodes
   ├╴constructor
   └╴testFFI()
      └╴error Revert (0x00000000000000000000000000000000000000000000000000000000000000566666692064697361626c65643a2072756e20616761696e2077697468202d2d66666920696620796f752077616e7420746f20616c6c6f7720746573747320746f2063616c6c2065787465726e616c207363726970747300000000000000000000) (src/broked.sol:24)
```

when I want to see this:

```
Failure: testFFI
  src/broked.sol:CheatCodes
   ├╴constructor
   └╴testFFI()
      └╴error Revert ("ffi disabled: run again with --ffi if you want to allow tests to call external scripts") (src/broked.sol:24)
```

Reply:
I tried making an explicit call to `finishFrame` as below, but I still get the same output:

```haskell
            let msg = "ffi disabled: run again with --ffi if you want to allow tests to call external scripts"
            in finishFrame . FrameReverted . ConcreteBuffer $ abiMethod "Error(string)" (AbiString . encodeUtf8 $ msg),
```

Reply:
The message should have been encoded as a string in a tuple, not just a string.